targetlibs/nrf5x/components/nfc/t2t_lib/hal_t2t/hal_nfc_t2t.c

```diff
@@ -615,45 +609,41 @@ void NFCT_IRQHandler(void)
         /* Take into account only number of whole bytes */
         uint32_t rx_data_size = ((NRF_NFCT->RXD.AMOUNT & NFCT_RXD_AMOUNT_RXDATABYTES_Msk) >>
                                  NFCT_RXD_AMOUNT_RXDATABYTES_Pos);
+
         /* Prevent integer underflow */
-        if(rx_data_size >= NFC_CRC_SIZE) rx_data_size -= NFC_CRC_SIZE; 
-        else rx_data_size = 0;
+        if(rx_data_size >= NFC_CRC_SIZE) rx_data_size -= NFC_CRC_SIZE;
 
         nrf_nfct_event_clear(&NRF_NFCT->EVENTS_RXFRAMEEND);
 
+        /* Use default FRAMEDELAY, for all cases but callback */
+        NRF_NFCT->FRAMEDELAYMAX = 0x1000UL; //302us, taken from datasheet
+
         /* Frame is garbage, wait for next frame reception */
-        if(rx_data_size == 0)
+        if((rx_data_size == 0) || (rx_data_size > NFC_BUFFER_SIZE))
         {
             NRF_NFCT->TASKS_ENABLERXDATA = 1;
-        }
-        /* Look for Tag 2 Type READ Command */
-        else if (m_nfc_rx_buffer[0] == T2T_READ_CMD)
+        } else
+        /* Indicate that SLP_REQ was received - this will cause FRAMEDELAYTIMEOUT error */
+        if(m_nfc_buffer[0] == NFC_SLP_REQ_CMD)
         {
-            if(m_nfc_lib_callback != NULL)
-            {
-                /* This callback should trigger transmission of READ Response */
-                m_nfc_lib_callback(m_nfc_lib_context,
-                                   HAL_NFC_EVENT_DATA_RECEIVED,
-                                   (void*)m_nfc_rx_buffer,
-                                   rx_data_size);
-            }
-        }
-        else
-        {
-            /* Indicate that SLP_REQ was received - this will cause FRAMEDELAYTIMEOUT error */
-            if(m_nfc_rx_buffer[0] == NFC_SLP_REQ_CMD)
-            {
-                m_slp_req_received = true;
-            }
-            /* Not a READ Command, so wait for next frame reception */
+            m_slp_req_received = true;
+
+            /* Wait for next frame reception */
             NRF_NFCT->TASKS_ENABLERXDATA = 1;
+
+        } else
+        if(m_nfc_lib_callback != NULL)
+        {
+            /* JS-has a high latency, set relaxed FRAMEDELAY and hope for a forgiving reader */

```

Comment:
I wonder whether the delay could be specified along with the callback? I'm just thinking in the case of the URLs, the delay is actually quite low again now - it's only JS that's an issue as you say.

Reply:
I've attached figure 119 from nRF52832_PS_v1.3.pdf
<img width="1069" alt="figure 119" src="https://user-images.githubusercontent.com/8958160/34252112-9707732c-e642-11e7-99d6-b5adfd22520f.png">
Max timeout is for a big part just a hint to the NFC peripheral’s state machine. It defines how long the peripheral should wait until it considers a missing tx command, after a rx callback, as an issue and fires an error interrupt. Therefore, as long as the rx callback initiates tx the max timeout has no effect.

However, the state machine almost always (depends on an internal state) waits after rx until max timeout is reached prior performing another rx. Hence problems may occur in cases where no response is required and sent. Quote from nRF52832_PS_v1.3.pdf: "This may happen even when the response is not required as per NFC Forum, NFC Digital Protocol Technical Specification. Any commands handled by the automatic collision resolution that don't involve a response being generated". In this such error cases the reader retries within a few ns. So keeping the timeout always at max prevents timely error recovery.

In conclusion, I would leave the implementation as it is as long as no bug reports for read errors come in. For the reasons above I also doubt it will ever hapen.

Reply:
Sounds great - thanks!